pandas/_libs/hashtable_func_helper.pxi.in

```diff
@@ -77,54 +77,76 @@ cdef build_count_table_{{dtype}}(const {{dtype}}_t[:] values,
 @cython.wraparound(False)
 @cython.boundscheck(False)
 {{if dtype == 'object'}}
-cpdef value_count_{{dtype}}(ndarray[{{dtype}}] values, bint dropna):
+cpdef value_count_{{dtype}}(ndarray[{{dtype}}] values, bint dropna, navalue=np.NaN):
 {{else}}
 cpdef value_count_{{dtype}}(const {{dtype}}_t[:] values, bint dropna):
 {{endif}}
     cdef:
         Py_ssize_t i = 0
+        Py_ssize_t n = len(values)
+        size_t unique_key_index = 0
+        size_t unique_key_count = 0
         kh_{{ttype}}_t *table
 
-        {{if dtype != 'object'}}
-        {{dtype}}_t[:] result_keys
-        int64_t[:] result_counts
-        {{endif}}
-
         # Don't use Py_ssize_t, since table.n_buckets is unsigned
         khiter_t k
+        bint is_null
+
+        {{c_type}} val
+
+        int ret = 0
+
+    # we track the order in which keys are first seen (GH39009),
+    # khash-map isn't insertion-ordered, thus:
+    #    table maps key to index_of_appearence
+    #    result_keys maps index_of_appearence to key
+    #    result_counts maps index_of_appearence to number of elements
 
+    result_keys = {{name}}Vector()
+    result_counts = Int64Vector()
     table = kh_init_{{ttype}}()
+
     {{if dtype == 'object'}}
-    build_count_table_{{dtype}}(values, table, 1)
+    kh_resize_{{ttype}}(table, n // 10)
+
+    for i in range(n):
+        val = values[i]
+        is_null = checknull(val)
+        if is_null:
+            val = navalue
+        if not is_null or not dropna:
```

Comment:
could be elif

Reply:
I don't see how it could work with elif (but I'm always struggeling with logic). I have rearanged the if and hope it is clearer now: https://github.com/pandas-dev/pandas/pull/39009/commits/230215fe6dccf5cbd2bee427eecf342139dfd037